lsp-completion.el

```diff
@@ -0,0 +1,553 @@
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+(defun lsp-completion--setup-company ()
+  "Setup company-mode."
+  (cond
+   ((and (functionp 'company-lsp)
+         (not lsp-completion-prefer-capf))
+    (progn
+      (company-mode 1)
+      (add-to-list 'company-backends 'company-lsp)
+      (setq-local company-backends (remove 'company-capf company-backends))))
+
+   ((and (fboundp 'company-mode) lsp-completion-enable)
+    (company-mode 1)
+    (add-to-list 'company-backends 'company-capf)))
+
+  (add-hook 'company-completion-started-hook
+            (lambda (&rest _)
+              (setq-local lsp-inhibit-lsp-hooks t))
+            nil
+            t)
+  (add-hook 'company-after-completion-hook
+            (lambda (&rest _)
+              (lsp-completion--capf-clear-cache)
+              (setq-local lsp-inhibit-lsp-hooks nil))
+            nil
+            t))
+
+(defun lsp-completion--clean-company ()
+  "Clean company-mode."
+  (remove-hook 'company-completion-started-hook
+               (lambda (&rest _)
+                 (setq-local lsp-inhibit-lsp-hooks t))
+               t)
+  (remove-hook 'company-after-completion-hook
+               (lambda (&rest _)
+                 (lsp-completion--capf-clear-cache)
+                 (setq-local lsp-inhibit-lsp-hooks nil))
+               t))
+
+(defun lsp-completion--enable ()
+  "Enable LSP completion support."
+  (when (and lsp-completion-enable
+             (lsp-feature? "textDocument/completion"))
+    (setq-local completion-at-point-functions nil)
+    (add-hook 'completion-at-point-functions #'lsp-completion-at-point nil t)
+    (setq-local completion-category-defaults
+                (add-to-list 'completion-category-defaults '(lsp-capf (styles basic))))
+    (lsp-completion-mode 1)))
+
+(defun lsp-completion--disable ()
+  "Disable LSP completion support."
+  (lsp-completion-mode -1)
+  (remove-hook 'completion-at-point-functions #'lsp-completion-at-point t)
+  (setq-local completion-category-defaults
+              (cl-remove 'lsp-capf completion-category-defaults :key #'car)))
+
+;;;###autoload
+(define-minor-mode lsp-completion-mode
+  "Toggle LSP completion support."
+  :group 'lsp-mode
+  :global nil
+  :lighter ""
+  (cond
+   (lsp-completion-mode
+    (add-hook 'lsp-configure-hook #'lsp-completion--enable nil t)
+    (add-hook 'lsp-unconfigure-hook #'lsp-completion--disable nil t)
+    (when (bound-and-true-p company-mode)
+      (lsp-completion--setup-company)))
+   (t
+    (remove-hook 'lsp-configure-hook #'lsp-completion--enable t)
+    (remove-hook 'lsp-unconfigure-hook #'lsp-completion--disable t)
+    (when (bound-and-true-p company-mode)
+      (lsp-completion--clean-company)))))
+
+;;;###autoload
+(add-hook 'lsp--auto-configure-hook #'lsp-completion--enable)
```

Comment:
This will not work since `lsp-completion--enable` is not autoloaded.
You should probably replace this with something like
``` elisp
;;;###autoload
(defun lsp-completion--auto-configure ()
  (setq lsp-completion-enable t)
  (cond
   ((and (functionp 'company-lsp)
         (not lsp-completion-prefer-capf))
    (progn
      (company-mode 1)
      (add-to-list 'company-backends 'company-lsp)
      (setq-local company-backends (remove 'company-capf company-backends))))
   ((and (fboundp 'company-mode) lsp-completion-enable)
    (company-mode 1)
    (add-to-list 'company-backends 'company-capf)))
  (lsp-completion-mode 1))

;;;###autoload
(add-hook 'lsp--auto-configure-hook #'lsp-completion--auto-configure)
```

Reply:
I just did this, but without the `(setq lsp-completion-enable t)`, why do we need it? I thought user would enable that